src/main/java/net/neoforged/moddevgradle/internal/ModDevPlugin.java

```diff
@@ -984,41 +995,58 @@ private static void addEclipseLaunchConfiguration(Project project,
+    record DataFileCollectionWrapper(DataFileCollection extension, Configuration configuration) {
+    }
 
+    private static DataFileCollectionWrapper dataFileConfiguration(Project project, String name, String description, String category) {
         var configuration = project.getConfigurations().create(name, spec -> {
             spec.setDescription(description);
             spec.setCanBeConsumed(false);
             spec.setCanBeResolved(true);
-            spec.withDependencies(dependencies -> dependencies.add(depFactory.create(collection.getFiles())));
+            spec.attributes(attributes -> attributes.attribute(Category.CATEGORY_ATTRIBUTE, project.getObjects().named(Category.class, category)));
         });
+
         var elementsConfiguration = project.getConfigurations().create(name + "Elements", spec -> {
             spec.setDescription("Published data files for " + name);
             spec.setCanBeConsumed(true);
             spec.setCanBeResolved(false);
-            spec.withDependencies(dependencies -> dependencies.add(depFactory.create(collection.getPublished())));
+            spec.attributes(attributes -> attributes.attribute(Category.CATEGORY_ATTRIBUTE, project.getObjects().named(Category.class, category)));
         });
 
-        // Set up the publishing conditionally
-        AdhocComponentWithVariants java = (AdhocComponentWithVariants) project.getComponents().getByName("java");
+        // Set up the variant publishing conditionally
+        var java = (AdhocComponentWithVariants) project.getComponents().getByName("java");
+        java.addVariantsFromConfiguration(elementsConfiguration, variant -> {
+            if (variant.getConfigurationVariant().getArtifacts().isEmpty()) {
+                variant.skip();
+            }
+        });
 
-        AtomicBoolean configured = new AtomicBoolean();
-        Runnable configurePublishing = () -> {
-            if (configured.compareAndSet(false, true)) {
-                java.addVariantsFromConfiguration(elementsConfiguration, variant -> {
+        var depFactory = project.getDependencyFactory();
+        Consumer<Object> publishCallback = new Consumer<>() {
+            ConfigurablePublishArtifact firstArtifact;
+            int artifactCount;
+
+            @Override
+            public void accept(Object artifactNotation) {
+                elementsConfiguration.getDependencies().add(depFactory.create(project.files(artifactNotation)));
+                project.getArtifacts().add(elementsConfiguration.getName(), artifactNotation, artifact -> {
+                    if (firstArtifact == null) {
+                        firstArtifact = artifact;
+                        artifact.setClassifier(category);
+                        artifactCount = 1;
+                    } else if (artifactCount == 1) {
+                        firstArtifact.setClassifier(category);
```

Comment:
Why change the classifier of only the first artifact? Shouldn't this change `artifact`?

Reply:
Ah yes, but `artifact` does not have its classifier set here. You are missing something like `artifact.setClassified(category + 2)`.